functions/api/category.js

```diff
@@ -0,0 +1,386 @@
+ */
+router.get("/:active/categories", async (request, response) => {
+    logger.info("Retrieving all active/inactive categories from firestore");
+    var status = JSON.parse(request.params.active.toLocaleLowerCase());
+    const categories = {
+        "categories": []
+    }
+    let categoryCollection = db.collection(constants.CATEGORY)
+        .where(constants.IS_ACTIVE, '==', status)
+    let snapshot = await categoryCollection.get()
+    snapshot.forEach(category => {
+        var categoryInfo = {}
+        var categoryData = category.data()
+        categoryInfo[constants.CATEGORY] = category.id,
+        categoryInfo[constants.DESCRIPTION] = categoryData.description,
+        categoryInfo[constants.IS_ACTIVE] = categoryData.isActive,
+        categoryInfo[constants.CREATED_DATE] = categoryData.createdDate.toDate(),
+        categoryInfo[constants.LAST_UPDATED_DATE] = categoryData.lastUpdatedDate.toDate()
+        categories.categories.push(categoryInfo);
+    })
+    categories[constants.TOTAL_CATEGORIES] = snapshot.size;
+    logger.debug('Returning categories to client.');
+    response.status(200).send(categories);
+});
+
+/**
+ * @description Route to add categories in Firestore
+ * @returns Created category
+ * @throws 400 if category already exists or if required params are missing
+ */
+router.post('/', async (request, response, next) => {
+    logger.info(`Creating category in firestore....`);
+    // Validate parameters
+    logger.debug('Validating params.')
+    const { error } = validateParams(request.body, constants.CATEGORY)
+    if (error) {
+        const err = new Error(error.details[0].message)
+        err.statusCode = 400
+        next(err)
+        return;
+    }
+
+    // If category already exists, return 400
+    var categoryName = request.body.category.toLocaleLowerCase()
+    logger.info(`Creating category ${categoryName} in firestore....`);
+    const doc = db.collection(constants.CATEGORY).doc(categoryName);
+    const category = await doc.get()
+    if (category.exists) {
+        const err = new Error(`The category ${categoryName} already exists. Please update if needed.`)
+        err.statusCode = 400
+        next(err)
+        return;
+    }
+
+    let data = {}
+    data[constants.DESCRIPTION] = request.body.description
+    data[constants.IS_ACTIVE] = true
+    data[constants.CREATED_DATE] = new Date()
+    data[constants.LAST_UPDATED_DATE] = new Date()
+    data[constants.PRODUCTS] = []
+    await db.collection(constants.CATEGORY).doc(categoryName).set(data)
+    logger.debug(`${categoryName} document Created`)
+    var result = {}
+    result[constants.CATEGORY] = categoryName
+    result[constants.DESCRIPTION] = data.description
+    response.status(200).send(result);    
+});
+
+/**
+ * @description Route to update description of category
+ * @returns  updated category
+ * @throws 400 if category does not exist or has wrong params
+ */
+router.put('/desc', async (request, response, next) => {
```

Comment:
It was just to have clean code. Avoiding if else loops.

Reply:
I am using this structure from my PUT apis's. Single api.
```
const { id } = req.body
    delete req.body.id
    await db.collection('roles').doc(id).update({
        ...req.body,
        lastUpdatedDate: Date.now()
    })
```
Do you see any issue with this modal, perhaps i missed?

Reply:
In this, how. will you verify the parameters, like not allowing blanks and all?

Reply:
something like this: like firstname if present, it validates if its a string
```
schema = joi.object().keys({
                id: joi.string().alphanum().min(28).max(30).required(),
                role: joi.string().regex(/^[a-z]{5,10}$/).required(),
                branch: joi.string().regex(/^[a-zA-Z]{5,30}$/).required(),
                firstName: joi.string().min(1).max(30),
                lastName: joi.string().min(1).max(30),
                contact: joi.string().length(10),
                isActive: joi.bool(),
                email: joi.string().email({ minDomainSegments: 2 }),
                createdDate: joi.date(),
                lastUpdatedDate: joi.date()
            })
```

Reply:
The problem with this is, incase if the user wants. to update firstName, it will not validate if blank is passed as .required is not set for firstName, similarly for other fields.

Reply:
in the above rules, firstName is optional. But if passed, It will validate against the conditions mentioned. min() and max() in this case.So, if i pass
```
{
  "id": "2zTYwPXLwPUYCH5n6hL1Ttu2aUD3",
	"role": "admin",
	"firstName": "",
  "branch": "sarjapur"
}
```
I get this error
```
{
  "status": 400,
  "message": "\"firstName\" is not allowed to be empty"
}
```

Reply:
Also, in the above case, the user will be able update createddate or lastupdatedDate. Have you handled that thing?